File: tests/e2e/test_examples.py
"""Tests that verify example outputs match .expect.txt golden files."""

from pathlib import Path

import pytest

from mermaid_ascii.api import render_dsl

EXAMPLES_DIR = Path(__file__).parent.parent.parent / "examples"


def find_example_pairs() -> list[tuple[str, Path, Path]]:
    """Find all .mm.md files that have a matching .expect.txt file."""
    pairs = []
    for mm_file in sorted(EXAMPLES_DIR.glob("*.mm.md")):
        expect_file = EXAMPLES_DIR / f"{mm_file.stem.replace('.mm', '')}.expect.txt"
        if expect_file.exists():
            name = mm_file.stem.replace(".mm", "")
            pairs.append((name, mm_file, expect_file))
    return pairs


EXAMPLE_PAIRS = find_example_pairs()


@pytest.fixture(autouse=True, scope="session")
def clean_out_files():
    """Delete all *.out.txt before and after the test session."""
    for f in EXAMPLES_DIR.rglob("*.out.txt"):
        f.unlink()
    yield
    for f in EXAMPLES_DIR.rglob("*.out.txt"):
        f.unlink()


@pytest.mark.parametrize("name,mm_file,expect_file", EXAMPLE_PAIRS, ids=[p[0] for p in EXAMPLE_PAIRS])
def test_example_matches_expect(name: str, mm_file: Path, expect_file: Path) -> None:
    """Render a .mm.md file and compare output against .expect.txt golden file."""
    src = mm_file.read_text()
    expected = expect_file.read_text()
    actual = render_dsl(src)
    assert actual == expected, f"Output for {name} differs from .expect.txt"

File: tests/e2e/test_examples.rs
//! Integration test: run ALL examples/*.mm.md through the full pipeline
//! and compare output against .expect.txt and .expect.svg golden files.

use mermaid_ascii::{render_dsl, render_svg_dsl};
use std::fs;
use std::path::Path;

#[test]
fn test_all_examples_txt() {
    let examples_dir = Path::new("examples");
    let mut tested = 0;
    let mut failures = Vec::new();

    for entry in fs::read_dir(examples_dir).expect("examples/ dir must exist") {
        let entry = entry.unwrap();
        let path = entry.path();
        if path.extension().map_or(true, |e| e != "md") {
            continue;
        }
        let name = path.file_stem().unwrap().to_string_lossy();
        if !name.ends_with(".mm") {
            continue;
        }
        let base = name.strip_suffix(".mm").unwrap();

        let expect_path = examples_dir.join(format!("{}.expect.txt", base));
        if !expect_path.exists() {
            continue;
        }

        let input = fs::read_to_string(&path).unwrap();
        let expected = fs::read_to_string(&expect_path).unwrap();
        let result = render_dsl(&input, true, 1, None).unwrap();

        if result.trim() != expected.trim() {
            failures.push(base.to_string());
        }
        tested += 1;
    }

    assert!(tested > 0, "no examples found to test");
    assert!(
        failures.is_empty(),
        "txt output mismatch for: {}",
        failures.join(", ")
    );
}

#[test]
fn test_all_examples_svg() {
    let examples_dir = Path::new("examples");
    let mut tested = 0;
    let mut failures = Vec::new();

    for entry in fs::read_dir(examples_dir).expect("examples/ dir must exist") {
        let entry = entry.unwrap();
        let path = entry.path();
        if path.extension().map_or(true, |e| e != "md") {
            continue;
        }
        let name = path.file_stem().unwrap().to_string_lossy();
        if !name.ends_with(".mm") {
            continue;
        }
        let base = name.strip_suffix(".mm").unwrap();

        let expect_path = examples_dir.join(format!("{}.expect.svg", base));
        if !expect_path.exists() {
            continue;
        }

        let input = fs::read_to_string(&path).unwrap();
        let expected = fs::read_to_string(&expect_path).unwrap();
        let result = render_svg_dsl(&input, 1, None).unwrap();

        if result.trim() != expected.trim() {
            failures.push(base.to_string());
        }
        tested += 1;
    }

    assert!(tested > 0, "no examples found to test");
    assert!(
        failures.is_empty(),
        "svg output mismatch for: {}",
        failures.join(", ")
    );
}
